git-branchless-revset/src/builtins.rs

```diff
@@ -178,8 +180,57 @@ fn fn_heads(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
 
 #[instrument]
 fn fn_branches(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
-    eval0(ctx, name, args)?;
-    Ok(ctx.dag.branch_commits.clone())
+    let pattern = match eval0_or_1_pattern(ctx, name, args)? {
+        Some(pattern) => pattern,
+        None => return Ok(ctx.dag.branch_commits.clone()),
+    };
+
+    let all_branch_commits: HashSet<_> = ctx
+        .dag
+        .commit_set_to_vec(&ctx.dag.branch_commits)
+        .map_err(EvalError::OtherError)?
+        .into_iter()
+        .collect();
+    let branch_commits = make_pattern_matcher(
+        ctx,
+        name,
+        args,
+        Box::new(move |repo: &Repo, commit: &Commit| {
+            if !all_branch_commits.contains(&commit.get_oid()) {
+                return Ok(false);
+            }
+
+            let branch_oid_to_names = match repo.get_references_snapshot() {
+                Ok(snapshot) => snapshot.branch_oid_to_names,
+                Err(err) => {
+                    warn!(?repo, ?err, "Could not get references snapshot for repo");
+                    return Ok(false);
+                }
+            };
+            let branches_at_commit = match branch_oid_to_names.get(&commit.get_oid()) {
+                Some(branches) => branches,
+                None => return Ok(false),
+            };
+
+            let result = branches_at_commit
+                .iter()
+                .filter_map(
+                    |branch_name| match CategorizedReferenceName::new(branch_name) {
+                        name @ CategorizedReferenceName::LocalBranch { .. } => {
+                            Some(name.render_suffix())
+                        }
```

Comment:
I recall using `render_suffix()` on a different feature, perhaps in `hide`, but I also note that in `submit` we use `name.remove_prefix()` in a similar situation. I couldn't tell the difference off hand, so I went w/ what I used previously here.

Reply:
Hmm... it looks like the only difference is that `remove_prefix` returns `Ok(...)` (unconditionally). They were probably different in the path, but we could remove `remove_prefix` now. (There was a time when we encoded references as `OsString`s instead of `String`s, but I dropped that feature, so some fallible code paths are no longer necessary).

Reply:
I went ahead and refactored these in this PR